### PR TITLE
Feature: XP & Character Level/XP Commands

### PR DESCRIPTION
**Pull request by kirmmin**
**Features**
- Grants XP to player
    - Saves XP totals to DB
- Executes Level Up
    - TBD: Update Stat packet needed
    - Bonus: Plays level up effect! (TY @ahoiahoi )
- Mega Bonus: Level up unlocks - replay feature added!! _"Boom! Welcome to club Bad Ass!"_

**Commands**
- `!character addxp <amount>`
   - Add the `<amount>` of XP to the player
- `!character level <value>`
   - Sets the player's level to the supplied `<value>`, and sets XP appropriately.
   - Note: Requires re-log currently. Will be resolved when StatUpdate packets are in.

### CODE REVIEW

File: Source/NexusForever.WorldServer/Game/Entity/Static/PlayerSaveMask.cs [hunk 1]
```diff
@@ -12,6 +12,7 @@ public enum PlayerSaveMask
         Level     = 0x0001,
         Location  = 0x0002,
         Path      = 0x0004,
-        Costume   = 0x0008
+        Costume   = 0x0008,
+        Xp        = 0x0016
```
Comment (ahoiahoi) [inline Source/NexusForever.WorldServer/Game/Entity/Static/PlayerSaveMask.cs]: 2x08 = 0x10 :)

Comment (kirmmin) [reply Source/NexusForever.WorldServer/Game/Entity/Static/PlayerSaveMask.cs]: oops! :D Thanks for pointing this out. It was working for me so it didn't register :p

File: Source/NexusForever.WorldServer/Network/Message/Model/ServerPlayerGrantXp.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+namespace NexusForever.WorldServer.Network.Message.Model
+{
+    [Message(GameMessageOpcode.ServerPlayerGrantXp, MessageDirection.Server)]
+    public class ServerPlayerGrantXp : IWritable
+    {
+        public uint TotalXpGained { get; set; }
+        public uint RestXpAmount { get; set; }
+        public uint SignatureXpAmount { get; set; }
+        public uint Unknown3 { get; set; }
+        public uint Unknown4 { get; set; }
```
Comment (Rawaho) [inline Source/NexusForever.WorldServer/Network/Message/Model/ServerPlayerGrantXp.cs]: `Unknown3` is called `reason`, uses the client enum `CodeEnumExpReason`
`Unknown4` is called `unitTarget` although I don't believe the client actually does anything with this value, the value is passed into the LUA event `ExperienceGained` but is never used

File: Source/NexusForever.WorldServer/Network/Message/Model/ServerPlayerGrantXp.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
+using System.Collections.Generic;
+using NexusForever.Shared.Network;
+using NexusForever.Shared.Network.Message;
+
+namespace NexusForever.WorldServer.Network.Message.Model
+{
+    [Message(GameMessageOpcode.ServerPlayerGrantXp, MessageDirection.Server)]
+    public class ServerPlayerGrantXp : IWritable
```
Comment (Rawaho) [inline Source/NexusForever.WorldServer/Network/Message/Model/ServerPlayerGrantXp.cs]: Any reason this was renamed?
The client calls the LUA event `ExperienceGained` in the native packet handler for 0x01AC, this is what it was originally named after

File: Source/NexusForever.WorldServer/Network/Message/Model/ClientReplayLevelRequest.cs [hunk 4]
```diff
@@ -0,0 +1,16 @@
+using NexusForever.Shared.Network;
+using NexusForever.Shared.Network.Message;
+
+namespace NexusForever.WorldServer.Network.Message.Model
+{
+    [Message(GameMessageOpcode.ClientReplayLevelRequest, MessageDirection.Client)]
+    public class ClientReplayLevelRequest : IReadable
```
Comment (Rawaho) [inline Source/NexusForever.WorldServer/Network/Message/Model/ClientReplayLevelRequest.cs]: The LUA function that invokes the native function to call this packet is called `ReplayLevelUp`, `ClientReplayLevelUp` would be a better name for this packet

File: Source/NexusForever.WorldServer/Game/Entity/Player.cs [hunk 5]
```diff
@@ -541,6 +558,93 @@ public void TeleportTo(WorldEntry entry, Vector3 vector, uint instanceId = 0u, u
...
+            PlayLevelUpEffect(newLevel);
+        }
+
+        public void PlayLevelUpEffect(byte level)
+        {
+            Spell4Entry spell4Entry = GameTableManager.Spell4.Entries
+                .Where(e => e.Spell4BaseIdBaseSpell == 53378 && e.TierIndex == level)
+                .FirstOrDefault();
+            SpellBaseInfo spellBaseInfo = GlobalSpellManager.GetSpellBaseInfo(spell4Entry.Spell4BaseIdBaseSpell);
+            SpellInfo spellInfo = spellBaseInfo.GetSpellInfo((byte)spell4Entry.TierIndex);
```
Comment (Rawaho) [inline Source/NexusForever.WorldServer/Game/Entity/Player.cs]: This is unnecessary, CastSpell has an overload that takes a baseSpellId and tier

File: Source/NexusForever.WorldServer/Network/Message/Model/ServerPlayerGrantXp.cs (hunk 3 above)
Comment (kirmmin) [reply Source/NexusForever.WorldServer/Network/Message/Model/ServerPlayerGrantXp.cs]: I will revert. I was trying to keep to the naming convention of most of our packets `(Direction)(MainClass)(Function)`.

File: Source/NexusForever.WorldServer/Game/Entity/Player.cs (hunk 5 above)
Comment (kirmmin) [reply Source/NexusForever.WorldServer/Game/Entity/Player.cs]: Now using the override you mentioned.

File: Source/NexusForever.WorldServer/Network/Message/Model/ClientReplayLevelRequest.cs (hunk 4 above)
Comment (kirmmin) [reply Source/NexusForever.WorldServer/Network/Message/Model/ClientReplayLevelRequest.cs]: Renamed packet as requested.

File: Source/NexusForever.WorldServer/Network/Message/Model/ServerPlayerGrantXp.cs (hunk 2 above)
Comment (kirmmin) [reply Source/NexusForever.WorldServer/Network/Message/Model/ServerPlayerGrantXp.cs]: Properties and packets have been renamed.
Added `ExpReason` enum, too.

File: Source/NexusForever.WorldServer/Game/Entity/Player.cs [hunk 6]
```diff
@@ -551,6 +569,114 @@ public void TeleportTo(WorldEntry entry, Vector3 vector, uint instanceId = 0u, u
...
         }
 
+        /// <summary>
+        /// Grants <see cref="Player"/> the supplied experience, handling level up if necessary.
+        /// </summary>
+        /// <param name="xp">Experience to grant</param>
+        /// <param name="reason"><see cref="ExpReason"/> for the experience grant</param>
+        public void GrantXp(uint xp, ExpReason reason = ExpReason.KillCreature)
+        {
+            uint maxLevel = 50;
```
Comment (Therzok) [inline Source/NexusForever.WorldServer/Game/Entity/Player.cs]:
```suggestion
            const uint maxLevel = 50;
```

**Comment by kirmmin** [comment]
This has been on my test server for a few months without issue. 

I know that #152 also has XP DB updates, so whichever one is merged second will need to update SQL & DB appropriately.

**Comment by kirmmin** [comment]
I decided to refactor this PR a little. I felt like there was a lot of overloading of the Player Entity class with this PR, so I created an `XpManager` class to manage it all.

I also added basic Rest Bonus XP support to this.

**Comment by kirmmin** [comment]
I went through and attempted my first conversion to a code first pattern. This one was dead simple, but I did need to modify the migration code after to camelcase the names. I'm not sure if I should be running a different command to build these things.

@Rawaho please review when you have time. It'd also be good if we can get some documentation on creating a DB migration for newcomers or, those like me, who just kinda stab around in the dark until something works :)

FYI, I used information from this page to help me make headway: https://www.entityframeworktutorial.net/efcore/entity-framework-core-migration.aspx
i.e. I used `Add-Migration CharacterXp -Context CharacterContext` in the Package Manager Console to create the migration.

**Pull request merged**